packages/protocol/contracts/L1/libs/LibL1Tokenomics.sol

```diff
@@ -55,126 +62,99 @@ library LibL1Tokenomics {
     )
         internal
-        view
-        returns (uint256 newFeeBase, uint256 reward, uint256 premiumRate)
+        pure
+        returns (
+            uint256 reward,
+            uint256 newProofTimeIssued,
+            uint64 newBaseFeeProof
+        )
     {
-        if (proposedAt > provenAt) revert L1_INVALID_PARAM();
-
-        if (state.lastVerifiedBlockId <= config.constantFeeRewardBlocks) {
-            reward = state.feeBase;
-            newFeeBase = state.feeBase;
-            // premiumRate = 0;
-        } else {
-            (newFeeBase, premiumRate) = getTimeAdjustedFee({
-                feeConfig: config.provingConfig,
-                feeBase: state.feeBase,
-                isProposal: false,
-                timeUsed: provenAt - proposedAt,
-                timeAverage: state.avgProofTime
-            });
-            reward = getSlotsAdjustedFee({
-                state: state,
-                config: config,
-                isProposal: false,
-                feeBase: newFeeBase
-            });
-        }
-        unchecked {
-            reward = (reward * (10000 - config.rewardBurnBips)) / 10000;
-        }
-    }
+        // To protect underflow
+        cumulativeProofTime = (cumulativeProofTime > proofTimeTarget)
+            ? cumulativeProofTime - proofTimeTarget
+            : uint256(0);
 
-    // Implement "Slot-availability Multipliers", see the whitepaper.
-    function getSlotsAdjustedFee(
-        TaikoData.State storage state,
-        TaikoData.Config memory config,
-        bool isProposal,
-        uint256 feeBase
-    ) internal view returns (uint256) {
-        unchecked {
-            // m is the `n'` in the whitepaper
-            uint256 m = 1000 *
-                config.maxNumProposedBlocks +
-                config.slotSmoothingFactor;
-            // n is the number of unverified blocks
-            uint256 n = 1000 *
-                (state.numBlocks - state.lastVerifiedBlockId - 1);
-            // k is `m − n + 1` or `m − n - 1`in the whitepaper
-            uint256 k = isProposal ? m - n - 1000 : m - n + 1000;
-            return (feeBase * (m - 1000) * m) / (m - n) / k;
-        }
-    }
+        cumulativeProofTime += actProofTime;
 
-    // Implement "Incentive Multipliers", see the whitepaper.
-    function getTimeAdjustedFee(
-        TaikoData.FeeConfig memory feeConfig,
-        uint256 feeBase,
-        bool isProposal,
-        uint256 timeUsed, // seconds
-        uint256 timeAverage // milliseconds
-    ) internal pure returns (uint256 newFeeBase, uint256 premiumRate) {
-        if (timeAverage == 0) {
-            return (feeBase, 0);
-        }
-        unchecked {
-            uint p = feeConfig.dampingFactorBips; // [0-10000]
-            uint a = timeAverage;
-            uint t = (timeUsed * 1000).min(a * 2); // millisconds
+        newBaseFeeProof = uint64(
+            baseFee(cumulativeProofTime, proofTimeTarget, quotient) /
+                SCALING_FACTOR_1E18
```

Comment:
If `baseFee` is in 10*18 fixed point (which it currently isn't, but okay :), I think it would be nice to keep it in that notation. If we do that we should use the full `uin256`, but seems like a safe thing to do for now.